### PR TITLE
Centralized logging for app services and Blazor API calls

### PR DESCRIPTION
**Pull request by dborgen2**
## What’s changing?

Added centralized logging in 3 places for issues that might arise during API calls or other requests.
1. Added logging in the RequestLoggingMiddleware, so that if an exception happens during the `await _next(context)` it will be caught and logged.
2. Added a centralized place to handle creating API responses in BaseController (CreateApiResponse), that will wrap the API request in a try/catch and log any exceptions, as well as safely returning an ApiResponse object with an error.
3. Added a centralized place in the Blazor code (PiipanApiService GetFromApi and PostToApi methods) that handle the ApiResponses, and gracefully creates an ApiResponse object with an error if the HTTP call fails.

Additional use cases not considered for this PR that we may want to investigate further in the future:
1. Overall error handling for Blazor client-side errors (such as null reference exception when rendering a page, etc).
2. Logging Blazor client side errors to Azure
3. Better handling of EAuth errors (because they potentially never hit our backend code)
4. Consider a general Exception Handling Middleware that would catch any rogue exceptions that fell through the cracks in a centralized place.

## Why?

Closes NAC-1868

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [x] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [x] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [x] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [x] Do the changes follow SOLID principles?
    - [x] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [x] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [x] For any new projects/solutions, have they been added to code coverage files?
    - [x] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [x] Do new capabilities (or changed capabilities) include logging?
    - [x] Is PII not logged?
- **General Error Handling Questions**
    - [x] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [x] Are errors/exceptions logged?
    - [x] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [x] End-to-End Tests provided (if code implements new user story)
    - [x] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs [hunk 1]
```diff
@@ -14,12 +15,19 @@ public RequestLoggingMiddleware(RequestDelegate next)
...
-            logger.LogInformation($"{claimsProvider.GetEmail(context.User)} {context.Request.Method} {context.Request.Path}");
-            await _next(context);
+            logger.LogInformation($"Request {context.TraceIdentifier}: {claimsProvider.GetEmail(context.User)} {context.Request.Method} {context.Request.Path}");
+            try
+            {
+                await _next(context);
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(ex, $"Request {context.TraceIdentifier} Exception: {ex.Message}");
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs:29]: Question: This exception is now getting swallowed. Is that an intentional change?

File: query-tool/src/Piipan.QueryTool/Controllers/DuplicateParticipantSearchController.cs [hunk 2]
```diff
@@ -32,96 +33,91 @@ public class DuplicateParticipantSearchController : BaseController<DuplicatePart
...
-            ApiResponse<OrchMatchResponseData> apiResponse = new();
-            await _webAppDataServiceProvider.Initialize();
-
-            // Only locations with length 2, which are states, can perform a search.
-            if (_webAppDataServiceProvider.Location.Length != 2 || _webAppDataServiceProvider.States?.Length != 1)
-            {
-                apiResponse.Errors.Add(new("", "Search performed with an invalid location"));
-            }
-            else if (ModelState.IsValid)
+            RegisterExceptionHandler<ArgumentException>(new()
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/Controllers/DuplicateParticipantSearchController.cs:43]: Question: Will this change with each request/search? i.e. is registration needed with each request?

File: shared/src/Piipan.Shared/Web/BaseController.cs [hunk 3]
```diff
@@ -49,5 +56,103 @@ protected List<(string Property, string ErrorMessage)> GetModelStateErrors(Model
...
+        /// </summary>
+        /// <typeparam name="TException"></typeparam>
+        /// <param name="exceptionHandler"></param>
+        /// <param name="statusCode"></param>
+        protected void RegisterExceptionHandler<TException>(ExceptionHandlerData<TException> exceptionHandlerData) where TException : Exception
+        {
+            Type exceptionType = typeof(TException);
+            if (exceptionHandlers.ContainsKey(exceptionType))
+            {
+                _logger.LogInformation($"Exception handler for type {exceptionType.Name} not applied. Exception already registered.");
```
[the remark this answers is not in the document]
Comment (dborgen2) [reply shared/src/Piipan.Shared/Web/BaseController.cs:71]: Hm... good question. I think if the same exception handler is getting registered twice, that is actually a programmer error. Since the registrations should happen before calling CreateApiResponse, there shouldn't be a case where you have multiple. If the exception needs to be handled inside the callback function from CreateApiResponse, the code should directly use HandleException (as noted in the comments) rather than registering it a second time.

File: shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs (hunk 1 above)
Comment (dborgen2) [reply shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs:29]: It was, but I'm wondering now if it should. Maybe it shouldn't, so that the redirect to the error page still happens. What are your thoughts?

File: query-tool/src/Piipan.QueryTool/Controllers/DuplicateParticipantSearchController.cs (hunk 2 above)
Comment (dborgen2) [reply query-tool/src/Piipan.QueryTool/Controllers/DuplicateParticipantSearchController.cs:43]: Yea, I thought about doing it at the constructor level vs request level. I think this gives it a little bit more customizability, if two Actions within the same controller would want to handle a specific error differently.

Comment (bgfoshay) [reply query-tool/src/Piipan.QueryTool/Controllers/DuplicateParticipantSearchController.cs:43]: 👍

File: shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs (hunk 1 above)
Comment (bgfoshay) [reply shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs:29]: It's an interesting question. My inclination would be to to say no. If an exception is being thrown (and not caught) by the API, I wouldn't necessarily place exception handling responsibility in the Logging middleware. Now, if we were registering exception handling middleware, e.g. something like ExceptionHandlingMiddleware, then swallowing the exceptions might be ok. That might be something we should consider in a future ticket

Comment (dborgen2) [reply shared/src/Piipan.Shared/Logging/RequestLoggingMiddleware.cs:29]: Sure, that makes a lot of sense. I'll throw it. I think mainly this is going to be for the initial page rendering (not API calls) since those should be well handled now, but I'll let Startup take care of it, and I'll add an ExceptionHandlingMiddleware to the list of additional use cases not covered by this work.

**Review by bgfoshay — APPROVED** [review]
LGTM- Thanks for answering all of my questions.
